File: src/claw_mem/context/confidence_gate.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""F2: ConfidenceGate + GateResult + ConfidenceLevel (v4.9.0 context control plane).

Four-dimensional confidence scoring:
  1. Vector  – search-result score as proxy for embedding distance
  2. Time    – TieredDecayEngine.classify() → tier-based decay signal
  3. Conflict– ConflictDetector batch cache; memory in any conflict scores low
  4. Frequency– tag-driven heuristic (e.g. missing tags → lower confidence)

When a dimension is unavailable (e.g. no TieredDecayEngine enabled), its weight
is redistributed proportionally across the remaining dimensions.
"""

from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional, Set


class ConfidenceLevel(Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class GateResult:
    """Individual evaluation result for a single memory."""

    memory_id: Optional[str]
    confidence_score: float
    confidence_level: ConfidenceLevel
    vector_score: float
    time_score: float
    conflict_score: float
    frequency_score: float
    reason: str = ""
    warning: Optional[str] = None


class ConfidenceGate:
    """Filters memories by four-dimensional confidence before context injection.

    Keeps HIGH (>= high_threshold) and MEDIUM memories; discards LOW
    (< low_threshold).  When a component is unavailable its weight is
    proportionally redistributed to the other dimensions.
    """

    # v4.9.0 defaults
    DEFAULT_HIGH = 0.7
    DEFAULT_LOW = 0.4
    DEFAULT_W_VECTOR = 0.4
    DEFAULT_W_TIME = 0.3
    DEFAULT_W_CONFLICT = 0.2
    DEFAULT_W_FREQUENCY = 0.1

    def __init__(
        self,
        manager: Any = None,
        high_threshold: float = DEFAULT_HIGH,
        low_threshold: float = DEFAULT_LOW,
        weight_vector: float = DEFAULT_W_VECTOR,
        weight_time: float = DEFAULT_W_TIME,
        weight_conflict: float = DEFAULT_W_CONFLICT,
        weight_frequency: float = DEFAULT_W_FREQUENCY,
    ):
        self.manager = manager
        self.high_threshold = high_threshold
        self.low_threshold = low_threshold
        self.weight_vector = weight_vector
        self.weight_time = weight_time
        self.weight_conflict = weight_conflict
        self.weight_frequency = weight_frequency

        # Cached conflict-id set (populated once per filter batch)
        self._conflict_cache: Optional[Set[str]] = None
        self._conflict_cache_filled = False

    # ── public API ─────────────────────────────────────────────────────

    def evaluate(self, memory: Dict[str, Any]) -> GateResult:
        """Score a single memory across four dimensions."""
        return self._score(memory)

    def evaluate_batch(self, memories: List[Dict]) -> List[GateResult]:
        """Score a batch of memories (builds conflict cache once)."""
        self._fill_conflict_cache()
        results = [self._score(m) for m in memories]
        self._reset_conflict_cache()
        return results

    def filter(self, memories: List[Dict]) -> List[Dict]:
        """Main entry point: evaluate batch then drop LOW-confidence items."""
        results = self.evaluate_batch(memories)
        kept: List[Dict] = []
        for i, result in enumerate(results):
            if result.confidence_level != ConfidenceLevel.LOW:
                kept.append(memories[i])
        return kept

    # ── scoring ────────────────────────────────────────────────────────

    def _score(self, memory: Dict[str, Any]) -> GateResult:
        """Composite score with weight redistribution on unavailable dims."""
        mem_id = memory.get("id")

        vs = self._compute_vector_score(memory)
        ts = self._compute_time_score(memory)
        cs = self._compute_conflict_score(memory)
        fs = self._compute_frequency_score(memory)

        # Determine which dimensions are available
        available = {
            "vector": vs is not None,
            "time": ts is not None,
            "conflict": cs is not None,
            "frequency": fs is not None,
        }
        # Gather effective weights, redistributing if unavailable
        eff_w = self._effective_weights(available)

        vs_val = vs if vs is not None else 0.0
        ts_val = ts if ts is not None else 0.0
        cs_val = cs if cs is not None else 0.0
        fs_val = fs if fs is not None else 0.0

        composite = (
            vs_val * eff_w["vector"]
            + ts_val * eff_w["time"]
            + cs_val * eff_w["conflict"]
            + fs_val * eff_w["frequency"]
        )

        # Map composite → level
        if composite >= self.high_threshold:
            level = ConfidenceLevel.HIGH
        elif composite >= self.low_threshold:
            level = ConfidenceLevel.MEDIUM
        else:
            level = ConfidenceLevel.LOW

        warnings = []
        if not available["time"]:
            warnings.append("time_score_unavailable")
        if not available["conflict"]:
            warnings.append("conflict_score_unavailable")

        reason = (
            f"vec={vs_val:.2f} t={ts_val:.2f} c={cs_val:.2f} f={fs_val:.2f}"
            f" → {composite:.2f} ({level.value})"
        )

        return GateResult(
            memory_id=mem_id,
            confidence_score=round(composite, 4),
            confidence_level=level,
            vector_score=round(vs_val, 4),
            time_score=round(ts_val, 4),
            conflict_score=round(cs_val, 4),
            frequency_score=round(fs_val, 4),
            reason=reason,
            warning="; ".join(warnings) if warnings else None,
        )

    def _effective_weights(self, available: Dict[str, bool]) -> Dict[str, float]:
        """If a dimension is unavailable, redistribute its weight proportionally."""
        w = {
            "vector": self.weight_vector,
            "time": self.weight_time,
            "conflict": self.weight_conflict,
            "frequency": self.weight_frequency,
        }
        unavailable_weight = sum(
            w[dim] for dim, ok in available.items() if not ok
        )
        available_weight = sum(
            w[dim] for dim, ok in available.items() if ok
        )
        if unavailable_weight <= 0 or available_weight <= 0:
            return w

        redistributed = unavailable_weight / available_weight
        for dim in w:
            if available[dim]:
                w[dim] += w[dim] * redistributed
            else:
                w[dim] = 0.0
        return w

    # ── dimension helpers ──────────────────────────────────────────────

    def _compute_vector_score(self, memory: Dict) -> Optional[float]:
        """Use search result `score` field as proxy for embedding distance.

        Returns None when the dimension should be treated as unavailable.
        """
        score = memory.get("score")
        if score is None:
            return None
        try:
            return float(score)
        except (TypeError, ValueError):
            return None

    def _compute_time_score(self, memory: Dict) -> Optional[float]:
        """Map TieredDecayEngine.classify() tier → numeric score.

        HOT=1.0, WARM=0.6, COLD=0.3.  Returns None if tiered decay is not
        enabled / available.
        """
        if self.manager is None:
            return None
        td = getattr(self.manager, "tiered_decay", None)
        if td is None:
            return None
        from ..decay.tiered_decay import TierLevel

        tier = td.classify(memory)
        return {TierLevel.HOT: 1.0, TierLevel.WARM: 0.6, TierLevel.COLD: 0.3}.get(
            tier, 0.3
        )

    def _compute_conflict_score(self, memory: Dict) -> Optional[float]:
        """Batch-conflict lookup: 0.3 if ID in any conflict, 1.0 if clean.

        Returns None when ConflictDetector is not available.
        """
        if self.manager is None:
            return None
        cd = getattr(self.manager, "conflict_detector", None)
        if cd is None:
            return None

        mem_id = memory.get("id")
        if mem_id is None:
            return 0.5  # unknown identity → neutral

        if self._conflict_cache is not None:
            return 0.3 if mem_id in self._conflict_cache else 1.0
        return 1.0  # no cache → assume clean (conservative)

    def _compute_frequency_score(self, memory: Dict) -> float:
        """Tag-driven heuristic: missing tags=0.5, has tags=0.8, critical=1.0."""
        tags = memory.get("tags") or memory.get("metadata", {}).get("tags") or []
        if not tags:
            return 0.5
        critical_keywords = ("critical", "important", "永久", "关键", "critical_rule")
        for t in tags:
            if isinstance(t, str) and any(kw in t.lower() for kw in critical_keywords):
                return 1.0
        return 0.8

    # ── conflict cache ─────────────────────────────────────────────────

    def _fill_conflict_cache(self) -> None:
        """Run ConflictDetector.detect_conflicts() once and cache conflicted IDs."""
        if self._conflict_cache_filled:
            return
        if self.manager is None:
            self._conflict_cache_filled = True
            return
        cd = getattr(self.manager, "conflict_detector", None)
        if cd is None:
            self._conflict_cache_filled = True
            return
        try:
            conflicts = cd.detect_conflicts()
            ids: Set[str] = set()
            for c in conflicts:
                ids.add(c.memory_id_a)
                ids.add(c.memory_id_b)
            self._conflict_cache = ids
        except Exception:
            self._conflict_cache = set()
        self._conflict_cache_filled = True

    def _reset_conflict_cache(self) -> None:
        self._conflict_cache = None
        self._conflict_cache_filled = False

File: src/context/confidence_gate.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * F2: ConfidenceGate + GateResult + ConfidenceLevel
 *
 * Four-dimensional confidence scoring:
 *   1. Vector  - search-result score as proxy for embedding distance
 *   2. Time    - tier-based decay signal
 *   3. Conflict- memory in any conflict scores low
 *   4. Frequency- tag-driven heuristic
 */

// ── Enums / Data classes ──────────────────────────────────────────────

export enum ConfidenceLevel {
  HIGH = "high",
  MEDIUM = "medium",
  LOW = "low",
}

export interface GateResult {
  memory_id?: string;
  confidence_score: number;
  confidence_level: ConfidenceLevel;
  vector_score: number;
  time_score: number;
  conflict_score: number;
  frequency_score: number;
  reason: string;
  warning?: string;
}

// ── ConfidenceGate ────────────────────────────────────────────────────

export class ConfidenceGate {
  static DEFAULT_HIGH = 0.7;
  static DEFAULT_LOW = 0.4;
  static DEFAULT_W_VECTOR = 0.4;
  static DEFAULT_W_TIME = 0.3;
  static DEFAULT_W_CONFLICT = 0.2;
  static DEFAULT_W_FREQUENCY = 0.1;

  manager: any;
  high_threshold: number;
  low_threshold: number;
  weight_vector: number;
  weight_time: number;
  weight_conflict: number;
  weight_frequency: number;

  private _conflict_cache: Set<string> | null = null;
  private _conflict_cache_filled = false;

  constructor(
    manager?: any,
    high_threshold: number = ConfidenceGate.DEFAULT_HIGH,
    low_threshold: number = ConfidenceGate.DEFAULT_LOW,
    weight_vector: number = ConfidenceGate.DEFAULT_W_VECTOR,
    weight_time: number = ConfidenceGate.DEFAULT_W_TIME,
    weight_conflict: number = ConfidenceGate.DEFAULT_W_CONFLICT,
    weight_frequency: number = ConfidenceGate.DEFAULT_W_FREQUENCY,
  ) {
    this.manager = manager;
    this.high_threshold = high_threshold;
    this.low_threshold = low_threshold;
    this.weight_vector = weight_vector;
    this.weight_time = weight_time;
    this.weight_conflict = weight_conflict;
    this.weight_frequency = weight_frequency;
  }

  // ── Public API ─────────────────────────────────────────────────────

  evaluate(memory: Record<string, unknown>): GateResult {
    return this._score(memory);
  }

  evaluate_batch(memories: Record<string, unknown>[]): GateResult[] {
    this._fill_conflict_cache();
    const results = memories.map((m) => this._score(m));
    this._reset_conflict_cache();
    return results;
  }

  filter(memories: Record<string, unknown>[]): Record<string, unknown>[] {
    const results = this.evaluate_batch(memories);
    const kept: Record<string, unknown>[] = [];
    for (let i = 0; i < results.length; i++) {
      if (results[i].confidence_level !== ConfidenceLevel.LOW) {
        kept.push(memories[i]);
      }
    }
    return kept;
  }

  // ── Scoring ────────────────────────────────────────────────────────

  private _score(memory: Record<string, unknown>): GateResult {
    const memId = memory.id as string | undefined;

    const vs = this._compute_vector_score(memory);
    const ts = this._compute_time_score(memory);
    const cs = this._compute_conflict_score(memory);
    const fs = this._compute_frequency_score(memory);

    const available: Record<string, boolean> = {
      vector: vs != null,
      time: ts != null,
      conflict: cs != null,
      frequency: fs != null,
    };

    const effW = this._effective_weights(available);

    const vsVal = vs ?? 0.0;
    const tsVal = ts ?? 0.0;
    const csVal = cs ?? 0.0;
    const fsVal = fs ?? 0.0;

    const composite =
      vsVal * effW.vector +
      tsVal * effW.time +
      csVal * effW.conflict +
      fsVal * effW.frequency;

    let level: ConfidenceLevel;
    if (composite >= this.high_threshold) {
      level = ConfidenceLevel.HIGH;
    } else if (composite >= this.low_threshold) {
      level = ConfidenceLevel.MEDIUM;
    } else {
      level = ConfidenceLevel.LOW;
    }

    const warnings: string[] = [];
    if (!available.time) warnings.push("time_score_unavailable");
    if (!available.conflict) warnings.push("conflict_score_unavailable");

    const reason = `vec=${vsVal.toFixed(2)} t=${tsVal.toFixed(2)} c=${csVal.toFixed(2)} f=${fsVal.toFixed(2)} \u2192 ${composite.toFixed(2)} (${level})`;

    return {
      memory_id: memId,
      confidence_score: Math.round(composite * 10000) / 10000,
      confidence_level: level,
      vector_score: Math.round(vsVal * 10000) / 10000,
      time_score: Math.round(tsVal * 10000) / 10000,
      conflict_score: Math.round(csVal * 10000) / 10000,
      frequency_score: Math.round(fsVal * 10000) / 10000,
      reason,
      warning: warnings.length > 0 ? warnings.join("; ") : undefined,
    };
  }

  private _effective_weights(available: Record<string, boolean>): Record<string, number> {
    const w: Record<string, number> = {
      vector: this.weight_vector,
      time: this.weight_time,
      conflict: this.weight_conflict,
      frequency: this.weight_frequency,
    };

    let unavailableWeight = 0;
    let availableWeight = 0;
    for (const dim of Object.keys(available)) {
      if (available[dim]) {
        availableWeight += w[dim];
      } else {
        unavailableWeight += w[dim];
      }
    }

    if (unavailableWeight <= 0 || availableWeight <= 0) return w;

    const redistributed = unavailableWeight / availableWeight;
    for (const dim of Object.keys(w)) {
      if (available[dim]) {
        w[dim] += w[dim] * redistributed;
      } else {
        w[dim] = 0.0;
      }
    }
    return w;
  }

  // ── Dimension helpers ──────────────────────────────────────────────

  private _compute_vector_score(memory: Record<string, unknown>): number | null {
    const score = memory.score;
    if (score == null) return null;
    const n = Number(score);
    return isNaN(n) ? null : n;
  }

  private _compute_time_score(memory: Record<string, unknown>): number | null {
    if (this.manager == null) return null;
    const td = (this.manager as any).tiered_decay;
    if (td == null) return null;

    // Map tier to score: HOT=1.0, WARM=0.6, COLD=0.3
    const tier = td.classify(memory);
    switch (tier) {
      case "HOT":
        return 1.0;
      case "WARM":
        return 0.6;
      case "COLD":
        return 0.3;
      default:
        return 0.3;
    }
  }

  private _compute_conflict_score(memory: Record<string, unknown>): number | null {
    if (this.manager == null) return null;
    const cd = (this.manager as any).conflict_detector;
    if (cd == null) return null;

    const memId = memory.id as string | undefined;
    if (memId == null) return 0.5;

    if (this._conflict_cache != null) {
      return this._conflict_cache.has(memId) ? 0.3 : 1.0;
    }
    return 1.0;
  }

  private _compute_frequency_score(memory: Record<string, unknown>): number {
    const tags: string[] =
      (memory.tags as string[]) ??
      ((memory.metadata as any)?.tags as string[]) ??
      [];
    if (tags.length === 0) return 0.5;

    const criticalKeywords = [
      "critical",
      "important",
      "\u6c38\u4e45",
      "\u5173\u952e",
      "critical_rule",
    ];
    for (const t of tags) {
      if (typeof t === "string") {
        const lower = t.toLowerCase();
        if (criticalKeywords.some((kw) => lower.includes(kw))) return 1.0;
      }
    }
    return 0.8;
  }

  // ── Conflict cache ─────────────────────────────────────────────────

  private _fill_conflict_cache(): void {
    if (this._conflict_cache_filled) return;
    if (this.manager == null) {
      this._conflict_cache_filled = true;
      return;
    }
    const cd = (this.manager as any).conflict_detector;
    if (cd == null) {
      this._conflict_cache_filled = true;
      return;
    }
    try {
      const conflicts = cd.detect_conflicts();
      const ids = new Set<string>();
      for (const c of conflicts) {
        ids.add(c.memory_id_a);
        ids.add(c.memory_id_b);
      }
      this._conflict_cache = ids;
    } catch {
      this._conflict_cache = new Set();
    }
    this._conflict_cache_filled = true;
  }

  private _reset_conflict_cache(): void {
    this._conflict_cache = null;
    this._conflict_cache_filled = false;
  }
}
